### ISSUE TITLE
Consider including Alpha and Beta labels to API versions

### ISSUE THREAD

**Issue by RandyLevensalor** [camaraproject/Commonalities#36]
**Problem description**
Pre 1.0.0 api are generally being considered to be either beta or alpha and are unstable.  With new releases that be break compatibility.  As mentioned during the recent governing committee call, all of these pre 1.0.0 APIs would be accesses as a v0 API.  And per the commonalities guideline, all v0 API calls should be backwards compatibility.

**Possible evolution**
Add an Alpha or Beta prefix to pre-released APIs and then increment the release version when backwards compatibility is broken.

So you could have v1Alpha1 and v2Alpha1 to indicate incompatible Alpha releases.

This could follow a similar api versioning rules to Kubernetes.  https://kubernetes.io/docs/reference/using-api/#api-versioning

**Alternative solution**
Move to new major releases when API backwards compatibility is broken.  However, this could lead to 1.0.0, 2.0.0 and later releases that are beta and alpha, which could be confusing.

**Additional context**
Many of the CAMARA APIs are undergoing several iterations before being becoming stable enough to be released and the versioning could be extended to align with this process.

**Comment by shilpa-padgaonkar** [comment]
There was such a discussion raised earlier https://github.com/camaraproject/WorkingGroups/discussions/137

Hopefully this time we get more votes :)

**Comment by Kevsy** [comment]
Just to note we should provide guidance on precedence, e.g. see [Semver](https://semver.org/) section 11.4

**Comment by eric-murray** [comment]
Some comments on this from me:
- This is (or should be) only an issue for draft or "pre-release" versions of an API when the major version number is 0. The issue will not apply for "production" releases, for then breaking changes will result in a major version increment. For pre-release, however, minor and even patch increments can be breaking. 
- We already have pre-releases with minor version 0, so changing the versioning system for pre-releases means either renumbering old versions or a break in the numbering system
- Alpha / beta numbering systems are really designed for software where extensive alpha / beta testing is required at scale. There is no software release associated with a CAMARA API update, so I don't expect APIs will remain in a draft state for extended periods, and certainly Vodafone would not go live with a pre-release version of the API. The main issue slowing "production" releases is that some common elements affecting multiple APIs are still not finalised, but that situation should improve over time.
- Finally, and most importantly, given that the published API specification must anyway be customised by the API provider to be usable (no API provider can say to a client "just go and download the API specification from the CAMARA github and then you are good to go"), then the basepath could also be modified as part of that process. Having an agreed basepath for pre-release versions is only an issue when multiple clients are testing multiple implementations.

So my counterproposal would be:
- Keep semantic versioning for pre-releases, perhaps formalising alpha / beta / rc extensions if agreed to be useful (e.g. `v0.9.0-rc`)
- Agree a "default" basepath for pre-releases that can be derived from the version number which changes with breaking changes (e.g. `/qod/v0-9-0/`)
- Emphasise to API implementors that they are free to change the basepath to anything they like for pre-release versions if it makes testing easier as they anyway need to give client implementors a customised API definition YAML

**Comment by pjhac** [comment]
Hello,

I would like to give my views on it.

First of all, I'm in favor of semver, and discussed it heavily with colleagues who have been using it for a while now.

I tried to summarize our discussion below, with the hope it might help on this topic. Some points might just look like facts, or repetitions of what was previously posted in this issue, and I hope it will help clarify a few things.

Again, this is a view built from experience, I'm not enforcing anything, just trying to show what works for us.

Semantic versioning on breaking changes:
- breaking changes imply a major version increment (that's the fundamental part)
- semver FAQ entry: https://semver.org/#if-even-the-tiniest-backward-incompatible-changes-to-the-public-api-require-a-major-version-bump-wont-i-end-up-at-version-4200-very-rapidly

What is a breaking change?
- turning an optional attribute into a mandatory one
- modifying or removing a mandatory attribute (breaks existing payloads)
- modifying or removing an endpoint (breaks existing queries)

What is a non-breaking change?
- turning a mandatory attribute into an optional one
- adding an optional attribute
- adding a new endpoint

How to avoid breaking changes?
- never change an endpoint name; instead, add a new one and mark the original one for deprecation in a minor release and remove it in a later major release (see semver FAQ entry: https://semver.org/#how-should-i-handle-deprecating-functionality)
- if possible, do the same for attributes

Endpoint path versioning:
- only use the major in all cases (even for pre-releases, see below)

Make the information available:
- provide an access to the API spec (via a link or dedicated endpoint)
- if possible, do the same to obtain the implemented API spec version

How to handle pre-releases (v0 major):
- it should be known by all parties that v0 will change constantly and is unstable (semver 2.0 section 4)
- only use the major in endpoint path, as for any other v1+ release; putting major.minor or even major.minor.patch in the path requires to keep track of all endpoints, or support dynamic spec implementation (to support all implementations going with different v0 subversions); this is going to quickly become unmanageable and is difficult for the client to manage as well -- unless one really wants to support multiple simultaneous deployments, either by code or concurrent services
- if there is a real need to support multiple versions, 2 consecutive major versions should be the maximum (this is by experience, not a convention or requirement)
- there must be communication to the users (through a dedicated endpoint, email and/or changes note); this is done for release versions, so why not for pre-release?

Dealing with alpha, beta, rc:
- follow semver 11.4: 1.0.0-alpha < 1.0.0-alpha.1 < 1.0.0-alpha.beta < 1.0.0-beta < 1.0.0-beta.2 < 1.0.0-beta.11 < 1.0.0-rc.1 < 1.0.0
- there is no need to use any other numbering, even for pre-releases

In short, and as indicated already, I'm not saying it is an absolute solution to all problems, but I believe that by following some basic rules, thinking and planning about what is/will be introduced in every release, it can turn into something manageable. One might consider the major versions may go very high, but when the impacts of the changes are thought through, major version increments to will tend to slow down as a better view of future changes takes shape, and this will be beneficial to all parties.

**Comment by rkandoi** [comment]
The current thinking as prosed in TSC ([PR to minutes to meeting](https://github.com/camaraproject/Governance/pull/103/files)) seems to be that we have a "stable version" which is released / under readiness tests and an "alpha" version which is "work in progress. In the Ericsson view, just having two stages is presumably enough and we do not need alpha/beta (..and more) stages. We also think that using the word "alpha" may be misleading and just using "wip" could be enough. (We have seen the wip approach already adopted in some subprojects).

One question though that remains is that we need mechanisms to mark specifications / endpoints as "release candidates" after specification is complete but before the readiness tests are carried out. This may also be desirable in the case error corrections on a "stable" specification are needed (but maybe semver addresses that naturally).

Added this comment here for completeness and the same has been mentioned in the "consolidation issue" [here](https://github.com/camaraproject/Commonalities/issues/67#issuecomment-1790210646).

**Comment by FabrizioMoggio** [comment]
Considering: https://github.com/camaraproject/Commonalities/issues/12#issuecomment-1709789616

in my understanding we can have code in 3 different stages:

1) code under development, not ready to be exposed to Developers
2) code with consolidated development and readiness checklist (https://github.com/camaraproject/WorkingGroups/blob/main/Commonalities/documentation/API-Readiness-Checklist.md) complete but for step 6. This code can be exposed to Developers with a note (Alpha or Beta) that it can be subjected to changes 
3) code completed and fully compliant with the readiness checklist. This code can be labelled as 1.0.0 (or more) and can be exposed to Developers as stable version. For this code the Developer expect more stability than an Alpha/Beta

in Commonalities Issue #12 we suggested to refer to 2) as Beta version.

According to the discussion above my understanding is that 2) **is an Alpha and Beta will not be used.** Is my understanding correct?
